### PR TITLE
Updates to numba extension type mapping and overload matching

### PR DESCRIPTION
**Pull request by maximusron**
- Type mapping no longer overwrites many-to-one nb_types to cpptypes
- Signature is  generated based on matched argument set and is joined to support function definitions with multiple arguments (see test https://github.com/maximusron/cppyy/blob/a669f79f6c502b83b3b262c69581f4fbc6e09b32/test/test_numba.py#L245-L268)
- Fixes the multiple templated argument issue in free templated functions [Non types is a WIP]
- Preliminary reference type support for builtins based on explicit typeref detection

### CODE REVIEW

**Comment by maximusron** [comment]
cc @sudo-panda  @vgvassilev

File: python/cppyy/numba_ext.py [hunk 1]
```diff
@@ -130,14 +142,59 @@ def get_call_type(self, context, args, kwds):
...
+        # TODO : This works for builtin reference type returns but not for objects, for now leaving it commented out
+        #  assuming no reference type tests are added. Use typeof_scope() and eventually merge this logic into cpp2numba
+        # if not str(type(return_type)).startswith("<class"):
+        #     if return_type.endswith("&"):
+        #         is_reference_return = True
+        #         return_type = return_type[:-1]
+        #     else:
+        #         is_reference_return = False
+
+        # print("Is reference:", is_reference_return)
```
Comment (sudo-panda) [inline python/cppyy/numba_ext.py:189]: It's better to leave this out and put this in another branch if it is something you are working on.

File: python/cppyy/numba_ext.py [hunk 2]
```diff
@@ -130,14 +142,59 @@ def get_call_type(self, context, args, kwds):
...
+        args_combinations = itertools.product(*args_mapping)
+
+        for arg_combo in args_combinations:
+            try:
+                if(len(arg_combo) > 1):
+                    signature = ", ".join(arg_combo)
+                elif(len(arg_combo) == 0):
+                    signature = ""
+                else:
+                    signature = arg_combo[0]
```
Comment (sudo-panda) [inline python/cppyy/numba_ext.py:163]: If I understand correctly, the error that we are trying to catch is not received through these statements, so these can be moved out of the try block.

**Review by sudo-panda — CHANGES_REQUESTED** [review]
Added some initial review comments.

File: python/cppyy/numba_ext.py [hunk 3]
```diff
@@ -48,14 +46,15 @@ class Qualified:
     'unsigned int'           : nb_types.uintc,
     'int32_t'                : nb_types.int32,
     'uint32_t'               : nb_types.uint32,
-    'int64_t'                : nb_types.int64,
```
[the remark this answers is not in the document]
Comment (maximusron) [reply python/cppyy/numba_ext.py]: had moved that when int64_t functions would give an unsigned long type not found back when the mapping depended on the dictionary order. Should'nt matter anymore, have restored the order in the latest commit

File: python/cppyy/numba_ext.py [hunk 4]
```diff
@@ -66,9 +65,14 @@ def cpp2numba(val):
...
-    _numba2cpp[value] = key
+    if value in _numba2cpp:
+        _numba2cpp[value].append(key)
+        continue
+    _numba2cpp[value] = [key]
+
+
 # prefer "int" in the case of intc over "int32_t"
-_numba2cpp[nb_types.intc] = 'int'
+# _numba2cpp[nb_types.intc].append('int')
```
[the remark this answers is not in the document]
Comment (maximusron) [reply python/cppyy/numba_ext.py]: done!

File: python/cppyy/numba_ext.py (hunk 2 above)
Comment (maximusron) [reply python/cppyy/numba_ext.py]: done!

**Comment by maximusron** [comment]
closing since [PR#177](https://github.com/wlav/cppyy/pull/177) builds up on this

**Pull request closed without merging**